### DISCUSSION TITLE
Migrating rule that checks EnvelopeFrom from SA to Rspamd

### DISCUSSION THREAD

**Discussion by kpapad904** [rspamd/rspamd#4042] [Q&A]
In SpamAssassin I have been using the following local rule:
```
header __LOCAL_BRAND_FROM From =~ /(brand|myBank)/i
header __LOCAL_BRAND_SENDER EnvelopeFrom =~ /brand\.tld/i
meta LOCAL_BRAND_PHISHING (__LOCAL_BRAND_FROM) && !(__LOCAL_BRAND_SENDER)
score LOCAL_BRAND_PHISHING 6.0
describe LOCAL_BRAND_PHISHING Fake mail from BRAND
```
to catch phishing mails where the From: field (in either the EmailAddress or the DisplayName parts) contains certain variations of "brand" name, but the EnvelopeFrom: isn't also from brand.tld. Like these:
```
From brand-comunication @t-online.de  Mon Sep 27 14:12:01 2021
From brand.mail-contact @magenta.de  Fri Nov  5 00:49:05 2021

From: "BRAND.myBank" <support @sharpshooter.rgr.jp>
From: "BRAND.myBank" <support @toshima-theatre.jp>
From: myBankBRAND <support @sharpshooter.rgr.jp>
From: BRAND <support @hayashi-dentalclinic.jp>
```
But my local SA rule doesn't work when transferred in verbatim to Rspamd (using the SA rules module) ... It flags legitimate emails as phishing.

PS: I also asked at the IRC channel, where cebka suggested that I changed 'EnvelopeFrom' to 'Envelope-From', however the official SA rules use the former syntax (e.g. in /usr/share/spamassassin/20_freemail.cf and /usr/share/spamassassin/72_active.cf )

Thank you in advance for your help and happy new year!
KP

**Comment by vstakhov** [comment]
Ok, after some thinking I concluded that my efforts to fix this shitty behaviour of SA are much more exhausting than your efforts to convert your rules to Rspamd using either selectors framework or something similar (even Lua). `EnvelopeFrom` support is there for some `freemail` plugin and nothing more. In general `src/plugins/lua/spamassassin.lua:210` onwards contains all things to adopt it to understand that `EnvelopeFrom` is not a real header name but a magic crap like `ALL`. But I will not do that for you, sorry.

**Reply by kpapad904** [reply]
Is there some readily available (native) Rspamd rule to catch this kind of phishing emails? Because they are very common.

Currently 80% of the phishing emails that reach my Inbox are of this kind, with fake `EmailAddress` and/or `DisplayName`, but they were not sent from `EnvelopeFrom`: brand.tld (note: these domains all use of strict SPF "-all" so the scammers can't fake  the `EnvelopeFrom`)

**Reply by vstakhov** [reply]
I need to check that but I'm always keen to include low FP and high spam catchment rules in Rspamd. SA compatibility seems to be less concern for me.

**Reply by kpapad904** [reply]
Can you convert my simple SA local rule to Rspamd _"using either selectors framework or something similar"_  as you wrote above? Or point me to some documentation and examples?

Such a rule probably won't be very useful if applied generally. That's why I wrote 4 individual rules, each regexp is searching for text specific to each one of my country's banks (which all use strict SPF "-all"). The regexp would be different for someone in France or UK.

**Comment by kpapad904** [comment]
A related question of mine is how Rspamd determines `EnvelopeFrom` and if I need to do some additional configuration.

In my current SA setup, since I'm using Postfix MTA, I defined `envelope_sender_header Return-Path` and `always_trust_envelope_sender 1` in local.cf.

According to SA docs there are two ways to determine `EnvelopeFrom`:
https://cwiki.apache.org/confluence/display/SPAMASSASSIN/EnvelopeSenderInHeaders
https://cwiki.apache.org/confluence/display/SPAMASSASSIN/EnvelopeSenderInReceived

**Comment by kpapad904** [comment]
An actual SA rule (one of several similar rules) I've been using to catch phishing emails supposedly from NBG:

```
header __LOCAL_NBG_FROM From =~ /([nν][bβв]g|national[ -]bank[ -]of[ -]greece|i-[bβв][aαåά][nν][kκк]|Εθνική[ -]Τράπεζα)/i
header __LOCAL_NBG_SENDER EnvelopeFrom =~ /(nbg.gr|nbgleasing.gr|ethnolease.gr|nbgsecurities.com)$/i
meta LOCAL_NBG_PHISHING (__LOCAL_NBG_FROM) && !(__LOCAL_NBG_SENDER)
score LOCAL_NBG_PHISHING 8.0
describe LOCAL_NBG_PHISHING Fake mail from NBG
```


And here is such a phishing email, which Rspamd did not catch:
```
Return-Path: <ucaki@kiel-canal.de>
Delivered-To: me@mydomain.tld
Received: from mail.mydomain.tld
	by hermes.mydomain.tld with LMTP
	id rg9VNlMHg2SgdgAAMmzRiA
	(envelope-from <ucaki@kiel-canal.de>)
	for <me@mydomain.tld>; Fri, 09 Jun 2023 14:04:51 +0300
Received: from smtprelay04.ispgateway.de (smtprelay04.ispgateway.de [80.67.18.16])
	by mail.mydomain.tld (Postfix) with ESMTPS id AF5A81943F
	for <me@mydomain.tld>; Fri,  9 Jun 2023 14:04:50 +0300 (EEST)
Received: from [193.233.18.114] (helo=smtp.kiel-canal.de)
	by smtprelay04.ispgateway.de with esmtpsa  (TLS1.2) tls TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
	(Exim 4.94.2)
	(envelope-from <ucaki@kiel-canal.de>)
	id 1q7ZvF-0004tY-Nq
	for me@mydomain.tld; Fri, 09 Jun 2023 13:04:49 +0200
Reply-To: National.Bank.of.Greece
X-Source-IP: 127.0.0.1
X-Sender-IP: 127.0.0.1
X-Mailer: Microsoft Office Outlook, Build 10.0.5610
X-MimeOLE: Produced By Microsoft MimeOLE V6.00.2800.1441
Received: Mayra Ask
Content-Type: multipart/mixed; boundary="===============0758334173995159883=="
MIME-Version: 1.0
To: me@mydomain.tld
From: National.Bank.of.Greece <ucaki@kiel-canal.de>
Subject: System Notification
Message-Id: <E1q7ZvF-0004tY-Nq@smtprelay04.ispgateway.de>
Date: Fri, 09 Jun 2023 13:04:49 +0200
X-Df-Sender: dWNha2lAa2llbC1jYW5hbC5kZQ==
X-Rspamd-Queue-Id: AF5A81943F
X-Spamd-Result: default: False [2.90 / 15.00];
	REPLYTO_UNPARSEABLE(1.00)[];
	CTYPE_MIXED_BOGUS(1.00)[];
	MIME_BASE64_TEXT_BOGUS(1.00)[];
	MIME_HTML_ONLY(0.20)[];
	R_SPF_ALLOW(-0.20)[+ip4:80.67.16.0/20];
	MIME_BASE64_TEXT(0.10)[];
	MIME_GOOD(-0.10)[multipart/mixed];
	RWL_MAILSPIKE_GOOD(-0.10)[80.67.18.16:from];
	FROM_EQ_ENVFROM(0.00)[];
	RCVD_VIA_SMTP_AUTH(0.00)[];
	ASN(0.00)[asn:8972, ipnet:80.67.16.0/20, country:DE];
	R_DKIM_NA(0.00)[];
	MIME_TRACE(0.00)[0:+,1:~];
	RCVD_IN_DNSWL_NONE(0.00)[80.67.18.16:from];
	RCPT_COUNT_ONE(0.00)[1];
	DMARC_NA(0.00)[kiel-canal.de: no valid DMARC record];
	ARC_NA(0.00)[];
	RCVD_TLS_ALL(0.00)[];
	FROM_HAS_DN(0.00)[];
	TO_DN_NONE(0.00)[];
	TO_MATCH_ENVRCPT_ALL(0.00)[];
	RCVD_COUNT_TWO(0.00)[2]
X-Rspamd-Action: no action
X-Rspamd-Server: mail.mydomain.tld
```
This particular bank phishing email was rather poorly designed, with bogus Reply-To etc.

Depending on their level of access on the hacked server, sometimes they'll spoof the entire `From:` field (both EmailAddress or the DisplayName parts e.g. `From: National.Bank.of.Greece <noreply@nbg.gr>`) and leave only the `Return-Path:` / EnvelopeFrom (which can't be spoofed if the bank's domain has a strict SPF record `-all`, which all greek banks have)
